### PR TITLE
[receiver/hostmetrics] add EntityState entity event as log record for host

### PR DESCRIPTION
**Pull request by povilasv**
**Description:** <Describe what has changed.>

Add EntityState event for host, implementation is similiar to k8sclusterreceiver. This change sends empty event, which can be used together with resourcedetection processor to add required attributes:
```
    receivers:
      hostmetrics:
        enabled: true
        metadata_collection_interval: "5m"

    processors:
      resourcedetection/env:
        detectors: ["system", "env"]
        timeout: 2s
        override: false
        system:
          resource_attributes:
            host.id:
              enabled: true
      resourcedetection/region:
        detectors: ["gcp", "ec2", "azure"]
        timeout: 2s
        override: true
      transform/entity-event:
        log_statements:
          - context: log
            statements:
              - set(attributes["otel.entity.attributes"]["otel.entity.id"]["host.id"], resource.attributes["host.id"])
              - set(attributes["otel.entity.attributes"]["host.name"], resource.attributes["host.name"])
              - set(attributes["otel.entity.attributes"]["host.type"], resource.attributes["host.type"])
              - set(attributes["otel.entity.attributes"]["host.image.id"], resource.attributes["host.image.id"])
              - set(attributes["otel.entity.attributes"]["host.image.name"], resource.attributes["host.image.name"])
          - context: resource
            statements:
              - keep_keys(attributes, [""])
    service:
      pipelines:
        logs/entity:
          exporters:
            - otlp
          processors:
            - k8sattributes
            - resourcedetection/env
            - resourcedetection/region
            - transform/entity-event
          receivers:
            - hostmetrics
```

**Link to tracking Issue:**

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33927

**Testing:** <Describe what testing was performed and which tests were added.>
- kind cluster
- unit tests

**Documentation:** <Describe the documentation added.>
- added docs

### CODE REVIEW

File: receiver/hostmetricsreceiver/config.go [hunk 1]
```diff
@@ -25,6 +26,13 @@ type Config struct {
 	Scrapers                       map[string]internal.Config `mapstructure:"-"`
 	// RootPath is the host's root directory (linux only).
 	RootPath string `mapstructure:"root_path"`
+
+	// Collection interval for metadata.
+	// Metadata of the particular entity in the cluster is collected when the entity changes.
```
Comment (tigrannajaryan) [inline receiver/hostmetricsreceiver/config.go:31]: Nit: I think the reference to "cluster" is not relevant to hosts.
```suggestion
	// Metadata of the particular entity is collected when the entity changes.
```

File: receiver/hostmetricsreceiver/config.go [hunk 2]
```diff
@@ -25,6 +26,13 @@ type Config struct {
...
+	// In addition metadata of all entities is collected periodically even if no changes happen.
+	// Setting the duration to 0 will disable periodic collection (however will not impact
+	// metadata collection on changes).
+	MetadataCollectionInterval time.Duration `mapstructure:"metadata_collection_interval"`
```
Comment (tigrannajaryan) [inline receiver/hostmetricsreceiver/config.go:35]: Is there a default value for `MetadataCollectionInterval` or you have to specify it explicitly in the config to enable periodic collection? I think I would prefer to default to periodic collection with some reasonably large interval (e.g. 5 min).

File: receiver/hostmetricsreceiver/receiver.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		settings: set,
+	}
+
+	return r
+}
+
+func (hmr *hostMetricsReceiver) sendEntityEvent(ctx context.Context) {
+	timestamp := pcommon.NewTimestampFromTime(time.Now())
+
+	out := metadataPkg.NewEntityEventsSlice()
```
Comment (tigrannajaryan) [inline receiver/hostmetricsreceiver/receiver.go:81]: In a follow up PR we should add "Interval" as a field to the Entity events and use `MetadataCollectionInterval` as the value. The "Interval" was proposed after the initial design and is now in [the OTEP](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#entitystate-event).

**Review by tigrannajaryan — APPROVED** [review]

**Review by tigrannajaryan — APPROVED** [review]

File: receiver/hostmetricsreceiver/README.md [hunk 4]
```diff
@@ -188,6 +190,9 @@ Currently, the hostmetrics receiver does not set any Resource attributes on the
 ```
 export OTEL_RESOURCE_ATTRIBUTES="service.name=<the name of your service>,service.namespace=<the namespace of your service>,service.instance.id=<uuid of the instance>"
 ```
+## Entity Events
+
+The hostmetrics receiver can send host entity event as log. See the [Entites Data Model](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#entity-events) spec for more details. By default, hostmetrics receiver sends periodic EntityState event every 5 minutes. You can change that by setting `metadata_collection_interval`.
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/README.md:195]: Please mention that this is experimental log-based entities that eventually will be replaced by the result of the OTEP

File: receiver/hostmetricsreceiver/factory.go [hunk 5]
```diff
@@ -48,7 +53,8 @@ func NewFactory() receiver.Factory {
 	return receiver.NewFactory(
 		metadata.Type,
 		createDefaultConfig,
-		receiver.WithMetrics(createMetricsReceiver, metadata.MetricsStability))
+		receiver.WithMetrics(createMetricsReceiver, metadata.MetricsStability),
+		receiver.WithLogs(createLogsReceiver, metadata.MetricsStability))
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/factory.go:57]:
```suggestion
		receiver.WithLogs(createLogsReceiver, metadata.LogsStability))
```

File: receiver/hostmetricsreceiver/factory.go [hunk 6]
```diff
@@ -81,14 +96,33 @@ func createMetricsReceiver(
...
 }
 
+func createLogsReceiver(
+	_ context.Context, set receiver.Settings, cfg component.Config, consumer consumer.Logs,
+) (receiver.Logs, error) {
+	oCfg := cfg.(*Config)
+
+	r := receivers.GetOrAdd(
+		cfg, func() component.Component {
+			return newHostMetricsReceiver(oCfg, &set)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/factory.go:120]: We cannot have different factories and reuse the result between metrics and logs. They are completely different objects `hostMetricsReceiver` and one created with `scraperhelper.NewScraperControllerReceiver` helper

Comment (dmitryax) [reply receiver/hostmetricsreceiver/factory.go:120]: I don't see a reason to use the shared receiver here

**Review by mx-psi — COMMENTED** [review]
Not a blocker but I find it a bit unusual that we would now have a component called 'hostmetrics' that emits telemetry that are not metrics (an event in this case).

**Comment by mx-psi** [comment]
There has not been any discussion about this on this repository, again, I am not opposed to the change happening, but I would like it to be discussed explicitly here on this repository, and not only on the Entities SIG. There are other stakeholders for this receiver apart from the Entities SIG (Collector SIG, System Semantic Conventions Working Group, most vendors have probably a stake on this receiver as well), so this discussion needs to happen with all parties involved.

**Comment by mx-psi** [comment]
Thanks! Adding that warning works for me for now

File: receiver/hostmetricsreceiver/receiver.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+
+	metadataPkg "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/experimentalmetricmetadata"
+)
+
+const entityType = "host"
+
+type hostMetricsReceiver struct {
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/receiver.go:21]: The name is a bit confusing. Can we call it `hostEntitiesReceiver`?

File: receiver/hostmetricsreceiver/factory.go [hunk 8]
```diff
@@ -89,6 +98,16 @@ func createMetricsReceiver(
...
 }
 
+func createLogsReceiver(
+	_ context.Context, set receiver.Settings, cfg component.Config, consumer consumer.Logs,
+) (receiver.Logs, error) {
+	oCfg := cfg.(*Config)
+
+	r := newHostMetricsReceiver(oCfg, &set)
+	r.nextLogs = consumer
+	return r, nil
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/factory.go:108]: No need to follow the pattern of a shared component
```suggestion
	return &hostMetricsReceiver{
		cfg:      cfg.(*Config),
		nextLogs: consumer,
		settings: set,
	}, nil
```

File: receiver/hostmetricsreceiver/README.md [hunk 9]
```diff
@@ -188,6 +190,9 @@ Currently, the hostmetrics receiver does not set any Resource attributes on the
 ```
 export OTEL_RESOURCE_ATTRIBUTES="service.name=<the name of your service>,service.namespace=<the namespace of your service>,service.instance.id=<uuid of the instance>"
 ```
+## Entity Events
+
+**Entity Events as logs are experimental** and might eventually be replaced by the result of the OTEP. The hostmetrics receiver can send the host entity event as a log. See the [Entites Data Model](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#entity-events) spec for more details. By default, the hostmetrics receiver sends periodic EntityState events every 5 minutes. You can change that by setting `metadata_collection_interval`. Entity Events as logs are experimental. The result of the OTEP might eventually replace that.
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/README.md:195]:
```suggestion
**Entity Events as logs are experimental** and might eventually be replaced by the result of [the OTEP](https://github.com/open-telemetry/oteps/blob/main/text/entities/0256-entities-data-model.md#entity-events). For now, the hostmetrics receiver can send the host entity event as a log records. By default, the hostmetrics receiver sends periodic EntityState events every 5 minutes. You can change that by setting `metadata_collection_interval`. Entity Events as logs are experimental. The result of the OTEP might eventually replace that.
```

File: receiver/hostmetricsreceiver/internal/metadata/generated_telemetry.go [hunk 10]
```diff
@@ -1,17 +0,0 @@
...
-	"go.opentelemetry.io/otel/trace"
-)
-
-func Meter(settings component.TelemetrySettings) metric.Meter {
-	return settings.MeterProvider.Meter("otelcol/hostmetricsreceiver")
-}
-
-func Tracer(settings component.TelemetrySettings) trace.Tracer {
-	return settings.TracerProvider.Tracer("otelcol/hostmetricsreceiver")
-}
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/metadata/generated_telemetry.go:17]: How this happen to be removed?

**Review by dmitryax — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Pull request merged**